toolchain/check/deduce.cpp

```diff
@@ -80,6 +80,28 @@ class DeductionWorklist {
     AddAll(context_.inst_blocks().Get(params), args, needs_substitution);
   }
 
+  auto AddAll(SemIR::StructTypeFieldsId params, SemIR::StructTypeFieldsId args,
+              bool needs_substitution) -> void {
+    const auto& param_fields = context_.struct_type_fields().Get(params);
+    const auto& arg_fields = context_.struct_type_fields().Get(args);
+    if (param_fields.size() != arg_fields.size()) {
+      // TODO: Decide whether to error on this or just treat the parameter list
+      // as non-deduced. For now we treat it as non-deduced.
+      return;
+    }
+    // Don't do deduction unless the names match in order.
+    // TODO: Support reordering of names.
+    for (auto [param, arg] : llvm::zip_equal(param_fields, arg_fields)) {
+      if (param.name_id != arg.name_id) {
+        return;
+      }
+    }
```

Comment:
How are each of these tested? I see where it's tested when the struct is consistent, but what about the various invalid cases?

Reply:
Added tests.